include/nbl/builtin/hlsl/subgroup/bitonic_sort.hlsl

```diff
@@ -0,0 +1,83 @@
+#ifndef NBL_BUILTIN_HLSL_SUBGROUP_BITONIC_SORT_INCLUDED
+#define NBL_BUILTIN_HLSL_SUBGROUP_BITONIC_SORT_INCLUDED
+#include "nbl/builtin/hlsl/bitonic_sort/common.hlsl"
+#include "nbl/builtin/hlsl/glsl_compat/subgroup_basic.hlsl"
+#include "nbl/builtin/hlsl/glsl_compat/subgroup_shuffle.hlsl"
+#include "nbl/builtin/hlsl/functional.hlsl"
+namespace nbl
+{
+namespace hlsl
+{
+namespace subgroup
+{
+
+template<typename KeyType, typename ValueType, typename Comparator = less<KeyType> >
+struct bitonic_sort_config
+{
+	using key_t = KeyType;
+	using value_t = ValueType;
+	using comparator_t = Comparator;
+};
+
+template<typename Config, class device_capabilities = void>
+struct bitonic_sort;
+
+template<typename KeyType, typename ValueType, typename Comparator, class device_capabilities>
+struct bitonic_sort<bitonic_sort_config<KeyType, ValueType, Comparator>, device_capabilities>
+{
+	using config_t = bitonic_sort_config<KeyType, ValueType, Comparator>;
+	using key_t = typename config_t::key_t;
+	using value_t = typename config_t::value_t;
+	using comparator_t = typename config_t::comparator_t;
+
+	static void mergeStage(uint32_t stage, bool bitonicAscending, uint32_t invocationID,
+		NBL_REF_ARG(pair<key_t, value_t>) loPair, NBL_REF_ARG(pair<key_t, value_t>) hiPair)
+	{
+		comparator_t comp;
+
+		[unroll]
+		for (uint32_t pass = 0; pass <= stage; pass++)
+		{
+			const uint32_t stride = 1u << (stage - pass); // Element stride
+			const uint32_t threadStride = stride >> 1;
+			if (threadStride == 0)
+			{
+				// Local compare and swap for stage 0
+				nbl::hlsl::bitonic_sort::compareSwap(bitonicAscending, loPair, hiPair, comp);
+			}
+			else
+			{
+				// Shuffle from partner using XOR
+				const key_t pLoKey = glsl::subgroupShuffleXor<key_t>(loPair.first, threadStride);
+				const value_t pLoVal = glsl::subgroupShuffleXor<value_t>(loPair.second, threadStride);
+				const key_t pHiKey = glsl::subgroupShuffleXor<key_t>(hiPair.first, threadStride);
+				const value_t pHiVal = glsl::subgroupShuffleXor<value_t>(hiPair.second, threadStride);
+
+				const pair<key_t, value_t> partnerLoPair = make_pair(pLoKey, pLoVal);
+				const pair<key_t, value_t> partnerHiPair = make_pair(pHiKey, pHiVal);
+
+				const bool isUpper = bool(invocationID & threadStride);
+				const bool takeLarger = isUpper == bitonicAscending;
+
+				nbl::hlsl::bitonic_sort::compareExchangeWithPartner(takeLarger, loPair, partnerLoPair, hiPair, partnerHiPair, comp);
```

Comment:
comments throughout code needed about this stuff https://github.com/Devsh-Graphics-Programming/Nabla/pull/940#discussion_r2411941883